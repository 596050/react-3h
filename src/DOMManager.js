class DOMManager {
  constructor() {
    this.mountedComponents = [];
    this.components = {};
    this.renderedCallbacks = [];
  }

  renderApp(rootSelector, Component, props) {
    if (!Component) {
      throw new Error('component must be provided');
    }
    if (!Component.isComponent) {
      throw new Error('main component must be a component extending Component');
    }
    const element = document.querySelector(rootSelector);
    if (!element) {
      throw new Error('Root element was not found');
    }
    this.root = element;
    this.app = new Component(props);
    this.refresh();
  }

  callAfterRender() {
    this.mountedComponents.forEach(instance => instance.baseComponentWasMounted());
    this.mountedComponents = [];

    this.renderedCallbacks.forEach(fn => fn());
    this.renderedCallbacks = [];

    this.removeUnmountedComponents();
  }

  isDetached(element) {
    if (!element) return true;
    let parent = element.parentElement;
    while (parent) {
      if (parent.tagName === 'BODY') return false;
      parent = parent.parentElement;
    }
    return true;
  }

  removeUnmountedComponents() {
    Object.keys(this.components).forEach((key) => {
      const component = this.components[key];
      if (this.isDetached(component.getRootElement())) {
        delete this.components[key];
        component.baseComponentWasUnmounted();
      }
    });
  }

  refresh() {
    this.root.innerHTML = this.app.renderComponent();
    this.callAfterRender();
  }

  renderComponentSnapshot(component, props = {}, children = '') {
    return this.renderComponent(component, 'snapshot', props, children, true);
  }

  renderComponent(Component, key, props = {}, children = '', isSnapshot = false) {
    if (!Component) {
      throw new Error('Component must be provided');
    }
    if (!key) {
      throw new Error('a unique key must be provided for the component ', Component);
    }
    const propsAndChildren = Object.assign({}, props, { children });
    let html = '';

    if (Component.isComponent) {
      let instance = this.components[key];
      if (!instance) {
        instance = new Component();
        if (!isSnapshot) {
          this.components[key] = instance;
          this.mountedComponents.push(instance);
        }
      }
      instance.setProps(propsAndChildren);
      html = instance.renderComponent(isSnapshot);
    } else {
      html = Component(propsAndChildren);
    }
    return html || '';
  }

  setAppProps(props) {
    this.app.setProps(props);
    this.refresh();
  }

  registerForRendered(callback) {
    this.renderedCallbacks.push(callback);
  }
}

const instance = new DOMManager();

export default instance;
