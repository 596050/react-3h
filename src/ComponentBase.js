import DOMManager from './DOMManager';
import { generateUID } from './utils';

export class ComponentBase {
  constructor(props = {}) {
    this.setProps(props);
    this.root = null;
    this.id = generateUID();
  }

  setProps(props) {
    this.props = Object.freeze(Object.assign({},props));
  }

  injectID(html) {
    let foundFirstTag = false;
    let spaceIndex = -1;
    for (let i = 0; i < html.length; i++ ) {
      const ch = html.charAt(i);
      if (foundFirstTag) {
        if (ch === ' ' || ch === '>') {
          spaceIndex = i;
          break;
        }
      } else if (ch === '<') {
        foundFirstTag = true;
      }
    }
    return `${html.substring(0,spaceIndex)} data-id="${this.id}"${html.substring(spaceIndex)}`;
  }

  renderComponent(isSnapshot = false) {
    if (!this.render) {
      throw new Error('Component must implement render function');
    }
    const html = this.render() || '';

    if (isSnapshot) return html;

    const htmlWithUniqueId = this.injectID(html);
    DOMManager.registerForRendered(() => {
      this.root = document.querySelector(`[data-id="${this.id}"]`);
      if (typeof this.componentWasRendered === 'function') {
        this.componentWasRendered();
      }
    });
    return htmlWithUniqueId;
  }

  _componentWasMounted() {
    this.root = document.querySelector(`[data-id="${this.id}"]`);
    if (typeof this.componentWasMounted === 'function') {
      this.componentWasMounted();
    }
  }

  _componentWasUnmounted() {
    this.root = null;
    if (typeof this.componentWasUnmounted === 'function') {
      this.componentWasUnmounted();
    }
  }

  setState(obj) {
    if (!this.state) {
      this.state = Object.freeze(Object.assign({},obj));
    } else {
      this.state = Object.freeze(Object.assign({},this.state,obj));
    }
    this.forceRender();
  }

  forceRender() {
    DOMManager.refresh();
  }

  getRootElement() {
    return this.root;
  }

  find(selector) {
    return this.getRootElement().querySelector(selector);
  }
}

ComponentBase.isComponent = true;
