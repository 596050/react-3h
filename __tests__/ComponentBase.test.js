import ComponentBase from '../src/ComponentBase';
import DOMManager from '../src/DOMManager';
import AppComponent from './helpers/AppComponent';

test('ComponentBase has a static property isComponent', () => {
  expect(ComponentBase.isComponent).toBe(true);
});

test('ComponentBase can be derived and rendered', () => {
  expect(DOMManager.renderComponentSnapshot(AppComponent)).toMatchSnapshot();
});