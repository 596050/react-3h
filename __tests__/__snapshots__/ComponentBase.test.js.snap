// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentBase can be derived and rendered 1`] = `"<div>App</div>"`;
